File: soc/bcm2710/gpio.c
/*
 * soc/bcm2710/gpio.c - BCM2710 GPIO Implementation
 *
 * Tutorial-OS: BCM2710 HAL Implementation
 *
 * The core register manipulation is UNCHANGED - we just added hal_gpio_*
 * functions that provide error codes and call the existing implementation.
 */

#include "hal/hal_gpio.h"
#include "hal/hal_timer.h"  /* For delay in pull config */
#include "bcm2710_regs.h"

/* =============================================================================
 * INTERNAL: HAL to BCM Function Mapping
 * =============================================================================
 * BCM uses non-sequential encoding for GPIO functions.
 */

static uint32_t hal_to_bcm_function(hal_gpio_function_t func)
{
    switch (func) {
        case HAL_GPIO_INPUT:    return BCM_GPIO_FUNC_INPUT;
        case HAL_GPIO_OUTPUT:   return BCM_GPIO_FUNC_OUTPUT;
        case HAL_GPIO_ALT0:     return BCM_GPIO_FUNC_ALT0;
        case HAL_GPIO_ALT1:     return BCM_GPIO_FUNC_ALT1;
        case HAL_GPIO_ALT2:     return BCM_GPIO_FUNC_ALT2;
        case HAL_GPIO_ALT3:     return BCM_GPIO_FUNC_ALT3;
        case HAL_GPIO_ALT4:     return BCM_GPIO_FUNC_ALT4;
        case HAL_GPIO_ALT5:     return BCM_GPIO_FUNC_ALT5;
        default:                return BCM_GPIO_FUNC_INPUT;
    }
}

static hal_gpio_function_t bcm_to_hal_function(uint32_t bcm_func)
{
    switch (bcm_func) {
        case BCM_GPIO_FUNC_INPUT:   return HAL_GPIO_INPUT;
        case BCM_GPIO_FUNC_OUTPUT:  return HAL_GPIO_OUTPUT;
        case BCM_GPIO_FUNC_ALT0:    return HAL_GPIO_ALT0;
        case BCM_GPIO_FUNC_ALT1:    return HAL_GPIO_ALT1;
        case BCM_GPIO_FUNC_ALT2:    return HAL_GPIO_ALT2;
        case BCM_GPIO_FUNC_ALT3:    return HAL_GPIO_ALT3;
        case BCM_GPIO_FUNC_ALT4:    return HAL_GPIO_ALT4;
        case BCM_GPIO_FUNC_ALT5:    return HAL_GPIO_ALT5;
        default:                    return HAL_GPIO_INPUT;
    }
}

/* =============================================================================
 * INTERNAL: Core GPIO Operations
 * =============================================================================
 */

/*
 * Set GPIO function - existing gpio_set_function() code
 */
static void bcm_gpio_set_function(uint32_t pin, uint32_t function)
{
    /*
     * Each GPFSEL register handles 10 pins, with 3 bits per pin.
     * Register offset = (pin / 10) * 4
     * Bit position = (pin % 10) * 3
     */
    uintptr_t reg_addr = BCM_GPIO_BASE + (pin / 10) * 4;
    uint32_t shift = (pin % 10) * 3;
    uint32_t mask = 0x7 << shift;

    uint32_t val = hal_mmio_read32(reg_addr);
    val = (val & ~mask) | (function << shift);
    hal_mmio_write32(reg_addr, val);
}

/*
 * Get GPIO function
 */
static uint32_t bcm_gpio_get_function(uint32_t pin)
{
    uintptr_t reg_addr = BCM_GPIO_BASE + (pin / 10) * 4;
    uint32_t shift = (pin % 10) * 3;

    uint32_t val = hal_mmio_read32(reg_addr);
    return (val >> shift) & 0x7;
}

/*
 * Set pull resistor - existing gpio_set_pull() code
 */
static void bcm_gpio_set_pull(uint32_t pin, uint32_t pull)
{
    /*
     * Pull resistor configuration requires a specific sequence:
     * 1. Write pull type to GPPUD
     * 2. Wait 150 cycles
     * 3. Write clock to GPPUDCLK0/1 for the pin
     * 4. Wait 150 cycles
     * 5. Clear GPPUD
     * 6. Clear GPPUDCLK0/1
     */
    hal_mmio_write32(BCM_GPPUD, pull);
    hal_delay_us(150);

    if (pin < 32) {
        hal_mmio_write32(BCM_GPPUDCLK0, 1 << pin);
    } else {
        hal_mmio_write32(BCM_GPPUDCLK1, 1 << (pin - 32));
    }
    hal_delay_us(150);

    hal_mmio_write32(BCM_GPPUD, 0);
    if (pin < 32) {
        hal_mmio_write32(BCM_GPPUDCLK0, 0);
    } else {
        hal_mmio_write32(BCM_GPPUDCLK1, 0);
    }
}

/* =============================================================================
 * HAL GPIO IMPLEMENTATION
 * =============================================================================
 */

hal_error_t hal_gpio_init(void)
{
    /* BCM2710 GPIO doesn't need special initialization */
    return HAL_SUCCESS;
}

hal_error_t hal_gpio_set_function(uint32_t pin, hal_gpio_function_t function)
{
    if (pin > BCM_GPIO_MAX_PIN) {
        return HAL_ERROR_GPIO_INVALID_PIN;
    }

    uint32_t bcm_func = hal_to_bcm_function(function);
    bcm_gpio_set_function(pin, bcm_func);

    return HAL_SUCCESS;
}

hal_error_t hal_gpio_get_function(uint32_t pin, hal_gpio_function_t *function)
{
    if (pin > BCM_GPIO_MAX_PIN) {
        return HAL_ERROR_GPIO_INVALID_PIN;
    }
    if (function == NULL) {
        return HAL_ERROR_NULL_PTR;
    }

    uint32_t bcm_func = bcm_gpio_get_function(pin);
    *function = bcm_to_hal_function(bcm_func);

    return HAL_SUCCESS;
}

hal_error_t hal_gpio_set_pull(uint32_t pin, hal_gpio_pull_t pull)
{
    if (pin > BCM_GPIO_MAX_PIN) {
        return HAL_ERROR_GPIO_INVALID_PIN;
    }

    uint32_t bcm_pull;
    switch (pull) {
        case HAL_GPIO_PULL_NONE: bcm_pull = BCM_GPIO_PULL_OFF; break;
        case HAL_GPIO_PULL_DOWN: bcm_pull = BCM_GPIO_PULL_DOWN; break;
        case HAL_GPIO_PULL_UP:   bcm_pull = BCM_GPIO_PULL_UP; break;
        default:                 bcm_pull = BCM_GPIO_PULL_OFF; break;
    }

    bcm_gpio_set_pull(pin, bcm_pull);
    return HAL_SUCCESS;
}

hal_error_t hal_gpio_set_high(uint32_t pin)
{
    if (pin > BCM_GPIO_MAX_PIN) {
        return HAL_ERROR_GPIO_INVALID_PIN;
    }

    if (pin < 32) {
        hal_mmio_write32(BCM_GPSET0, 1 << pin);
    } else {
        hal_mmio_write32(BCM_GPSET1, 1 << (pin - 32));
    }

    return HAL_SUCCESS;
}

hal_error_t hal_gpio_set_low(uint32_t pin)
{
    if (pin > BCM_GPIO_MAX_PIN) {
        return HAL_ERROR_GPIO_INVALID_PIN;
    }

    if (pin < 32) {
        hal_mmio_write32(BCM_GPCLR0, 1 << pin);
    } else {
        hal_mmio_write32(BCM_GPCLR1, 1 << (pin - 32));
    }

    return HAL_SUCCESS;
}

hal_error_t hal_gpio_write(uint32_t pin, bool value)
{
    return value ? hal_gpio_set_high(pin) : hal_gpio_set_low(pin);
}

bool hal_gpio_read(uint32_t pin)
{
    if (pin > BCM_GPIO_MAX_PIN) {
        return false;
    }

    uint32_t val;
    if (pin < 32) {
        val = hal_mmio_read32(BCM_GPLEV0);
        return (val & (1 << pin)) != 0;
    } else {
        val = hal_mmio_read32(BCM_GPLEV1);
        return (val & (1 << (pin - 32))) != 0;
    }
}

hal_error_t hal_gpio_toggle(uint32_t pin)
{
    bool current = hal_gpio_read(pin);
    return hal_gpio_write(pin, !current);
}

bool hal_gpio_is_valid(uint32_t pin)
{
    return pin <= BCM_GPIO_MAX_PIN;
}

uint32_t hal_gpio_max_pin(void)
{
    return BCM_GPIO_MAX_PIN;
}

/* =============================================================================
 * BULK OPERATIONS
 * =============================================================================
 */

hal_error_t hal_gpio_set_mask(uint32_t mask, uint32_t bank)
{
    if (bank == 0) {
        hal_mmio_write32(BCM_GPSET0, mask);
    } else if (bank == 1) {
        hal_mmio_write32(BCM_GPSET1, mask);
    } else {
        return HAL_ERROR_INVALID_ARG;
    }
    return HAL_SUCCESS;
}

hal_error_t hal_gpio_clear_mask(uint32_t mask, uint32_t bank)
{
    if (bank == 0) {
        hal_mmio_write32(BCM_GPCLR0, mask);
    } else if (bank == 1) {
        hal_mmio_write32(BCM_GPCLR1, mask);
    } else {
        return HAL_ERROR_INVALID_ARG;
    }
    return HAL_SUCCESS;
}

uint32_t hal_gpio_read_mask(uint32_t bank)
{
    if (bank == 0) {
        return hal_mmio_read32(BCM_GPLEV0);
    } else if (bank == 1) {
        return hal_mmio_read32(BCM_GPLEV1);
    }
    return 0;
}

/* =============================================================================
 * PERIPHERAL CONFIGURATION
 * =============================================================================
 * These exist in gpio_configure_for_*() functions.
 */

hal_error_t hal_gpio_configure_dpi(void)
{
    /*
     * Configure GPIO 0-27 for DPI display output (24-bit color)
     *
     * IMPORTANT: GPIO 18 and 19 are EXCLUDED from DPI configuration
     * because they are used for PWM audio output on the GPi Case 2W.
     *
     * DPI uses ALT2 function:
     *   GPIO 0-7:   Blue/Green data bits
     *   GPIO 8-15:  Green/Red data bits
     *   GPIO 16-17: HSYNC, VSYNC
     *   GPIO 18-19: RESERVED FOR PWM AUDIO (not configured here!)
     *   GPIO 20-27: Red channel / control signals
     */

    const uint32_t ALT2 = BCM_GPIO_FUNC_ALT2;

    /* GPFSEL0: GPIO 0-9 - All ALT2 for DPI */
    uint32_t gpfsel0_val = (ALT2 << 0)  | (ALT2 << 3)  | (ALT2 << 6)  | (ALT2 << 9)
                         | (ALT2 << 12) | (ALT2 << 15) | (ALT2 << 18) | (ALT2 << 21)
                         | (ALT2 << 24) | (ALT2 << 27);

    /* GPFSEL1: GPIO 10-17 ALT2, GPIO 18-19 left as input (for PWM audio) */
    uint32_t gpfsel1_val = (ALT2 << 0)  | (ALT2 << 3)  | (ALT2 << 6)  | (ALT2 << 9)
                         | (ALT2 << 12) | (ALT2 << 15) | (ALT2 << 18) | (ALT2 << 21);
    /* NOTE: No (ALT2 << 24) or (ALT2 << 27) - GPIO 18/19 reserved for audio! */

    /* GPFSEL2: GPIO 20-27 - All ALT2 for DPI */
    uint32_t gpfsel2_val = (ALT2 << 0)  | (ALT2 << 3)  | (ALT2 << 6)  | (ALT2 << 9)
                         | (ALT2 << 12) | (ALT2 << 15) | (ALT2 << 18) | (ALT2 << 21);

    hal_mmio_write32(BCM_GPFSEL0, gpfsel0_val);
    hal_mmio_write32(BCM_GPFSEL1, gpfsel1_val);
    hal_mmio_write32(BCM_GPFSEL2, gpfsel2_val);

    /* Memory barrier to ensure GPIO configuration is complete */
    HAL_DMB();

    /* Disable pull-up/down on DPI pins (GPIO 0-17 and 20-27) */
    hal_mmio_write32(BCM_GPPUD, BCM_GPIO_PULL_OFF);
    hal_delay_us(150);
    hal_mmio_write32(BCM_GPPUDCLK0, 0x0FF3FFFF);  /* Bits 0-17 and 20-27 */
    hal_delay_us(150);
    hal_mmio_write32(BCM_GPPUD, 0);
    hal_mmio_write32(BCM_GPPUDCLK0, 0);

    return HAL_SUCCESS;
}

hal_error_t hal_gpio_configure_hdmi(void)
{
    /* HDMI doesn't require GPIO configuration on BCM2710 */
    return HAL_SUCCESS;
}

hal_error_t hal_gpio_configure_sdcard(void)
{
    /*
     * Configure GPIO 48-53 for SDHOST controller (ALT0 function)
     *
     * Pin mapping:
     *   GPIO 48: SD_CLK
     *   GPIO 49: SD_CMD
     *   GPIO 50: SD_DAT0
     *   GPIO 51: SD_DAT1
     *   GPIO 52: SD_DAT2
     *   GPIO 53: SD_DAT3
     */

    const uint32_t ALT0 = BCM_GPIO_FUNC_ALT0;

    /* GPIO 48-49 are in GPFSEL4 (bits 24-29) */
    uint32_t gpfsel4 = hal_mmio_read32(BCM_GPFSEL4);
    gpfsel4 = (gpfsel4 & 0xC0FFFFFF) | (ALT0 << 24) | (ALT0 << 27);
    hal_mmio_write32(BCM_GPFSEL4, gpfsel4);

    /* GPIO 50-53 are in GPFSEL5 (bits 0-11) */
    uint32_t gpfsel5 = hal_mmio_read32(BCM_GPFSEL5);
    gpfsel5 = (gpfsel5 & 0xFFFFF000)
            | (ALT0 << 0)    /* GPIO 50 */
            | (ALT0 << 3)    /* GPIO 51 */
            | (ALT0 << 6)    /* GPIO 52 */
            | (ALT0 << 9);   /* GPIO 53 */
    hal_mmio_write32(BCM_GPFSEL5, gpfsel5);

    /* Enable pull-ups on data and command lines (GPIO 49-53) */
    hal_mmio_write32(BCM_GPPUD, BCM_GPIO_PULL_UP);
    hal_delay_us(150);
    hal_mmio_write32(BCM_GPPUDCLK1, (1 << 17) | (1 << 18) | (1 << 19) | (1 << 20) | (1 << 21));
    hal_delay_us(150);
    hal_mmio_write32(BCM_GPPUD, 0);
    hal_mmio_write32(BCM_GPPUDCLK1, 0);

    return HAL_SUCCESS;
}

hal_error_t hal_gpio_configure_audio(void)
{
    /*
     * Configure GPIO 18-19 for PWM audio (ALT5 function)
     *
     * Pin mapping:
     *   GPIO 18: PWM0 -> Left channel
     *   GPIO 19: PWM1 -> Right channel
     */

    const uint32_t ALT5 = BCM_GPIO_FUNC_ALT5;

    /* Read current GPFSEL1, modify only GPIO 18-19 (bits 24-29) */
    uint32_t gpfsel1 = hal_mmio_read32(BCM_GPFSEL1);
    gpfsel1 = (gpfsel1 & 0xC0FFFFFF)
            | (ALT5 << 24)    /* GPIO 18 = ALT5 (PWM0) */
            | (ALT5 << 27);   /* GPIO 19 = ALT5 (PWM1) */
    hal_mmio_write32(BCM_GPFSEL1, gpfsel1);

    /* No pull resistors needed for PWM output */
    hal_mmio_write32(BCM_GPPUD, BCM_GPIO_PULL_OFF);
    hal_delay_us(150);
    hal_mmio_write32(BCM_GPPUDCLK0, (1 << 18) | (1 << 19));
    hal_delay_us(150);
    hal_mmio_write32(BCM_GPPUD, 0);
    hal_mmio_write32(BCM_GPPUDCLK0, 0);

    return HAL_SUCCESS;
}

hal_error_t hal_gpio_configure_uart(uint32_t uart_num)
{
    if (uart_num == 0) {
        /* UART0 (PL011) uses GPIO 14 (TX) and 15 (RX) - ALT0 */
        bcm_gpio_set_function(14, BCM_GPIO_FUNC_ALT0);
        bcm_gpio_set_function(15, BCM_GPIO_FUNC_ALT0);
        bcm_gpio_set_pull(14, BCM_GPIO_PULL_OFF);
        bcm_gpio_set_pull(15, BCM_GPIO_PULL_UP);
    } else if (uart_num == 1) {
        /* UART1 (Mini UART) uses GPIO 14 (TX) and 15 (RX) - ALT5 */
        bcm_gpio_set_function(14, BCM_GPIO_FUNC_ALT5);
        bcm_gpio_set_function(15, BCM_GPIO_FUNC_ALT5);
        bcm_gpio_set_pull(14, BCM_GPIO_PULL_OFF);
        bcm_gpio_set_pull(15, BCM_GPIO_PULL_UP);
    } else {
        return HAL_ERROR_INVALID_ARG;
    }

    return HAL_SUCCESS;
}

hal_error_t hal_gpio_configure_safe_shutdown(void)
{
    /* GPi Case 2W uses GPIO 26 for safe shutdown button */
    bcm_gpio_set_function(26, BCM_GPIO_FUNC_INPUT);
    bcm_gpio_set_pull(26, BCM_GPIO_PULL_UP);
    return HAL_SUCCESS;
}

bool hal_gpio_safe_shutdown_triggered(void)
{
    /* GPIO 26 is active low on GPi Case */
    return !hal_gpio_read(26);
}

File: soc/bcm2710/src/gpio.rs
//! BCM2710 GPIO implementation.
//!
//! 54 GPIO pins (0–53), each with a 3-bit function select in the GPFSEL
//! registers.  BCM uses a non-sequential encoding for alternate functions
//! (ALT0=4, ALT1=5, ALT2=6, ALT3=7, ALT4=3, ALT5=2).  The mapping
//! tables in this module translate between the HAL's sequential enum and
//! BCM's hardware encoding.

use hal::gpio::{Function, Gpio, Peripheral, Pull};
use hal::types::{HalError, HalResult};
use crate::regs;

/// BCM2710 GPIO controller.
pub struct Bcm2710Gpio;

// ============================================================================
// HAL ↔ BCM function code mapping
// ============================================================================

fn hal_to_bcm(func: Function) -> u32 {
    match func {
        Function::Input  => regs::gpio_func::INPUT,
        Function::Output => regs::gpio_func::OUTPUT,
        Function::Alt0   => regs::gpio_func::ALT0,
        Function::Alt1   => regs::gpio_func::ALT1,
        Function::Alt2   => regs::gpio_func::ALT2,
        Function::Alt3   => regs::gpio_func::ALT3,
        Function::Alt4   => regs::gpio_func::ALT4,
        Function::Alt5   => regs::gpio_func::ALT5,
    }
}

fn bcm_to_hal(bcm: u32) -> Function {
    match bcm {
        0 => Function::Input,
        1 => Function::Output,
        4 => Function::Alt0,
        5 => Function::Alt1,
        6 => Function::Alt2,
        7 => Function::Alt3,
        3 => Function::Alt4,
        2 => Function::Alt5,
        _ => Function::Input,
    }
}

// ============================================================================
// Internal register-level operations
// ============================================================================

fn set_function_raw(pin: u32, func: u32) {
    let reg = regs::GPIO_BASE + (pin / 10) as usize * 4;
    let shift = (pin % 10) * 3;
    let mask = 0x7 << shift;
    let val = unsafe { common::mmio::read32(reg) };
    unsafe { common::mmio::write32(reg, (val & !mask) | (func << shift)) };
}

fn get_function_raw(pin: u32) -> u32 {
    let reg = regs::GPIO_BASE + (pin / 10) as usize * 4;
    let shift = (pin % 10) * 3;
    (unsafe { common::mmio::read32(reg) } >> shift) & 0x7
}

fn set_pull_raw(pin: u32, pull: u32, timer: &crate::timer::Bcm2710Timer) {
    use hal::timer::Timer;

    // BCM2710 pull configuration requires a specific dance:
    // 1. Write pull type to GPPUD
    // 2. Wait 150 cycles (~150 µs at 1 MHz timer)
    // 3. Write clock to GPPUDCLK for the pin
    // 4. Wait 150 cycles
    // 5. Clear GPPUD
    // 6. Clear GPPUDCLK
    unsafe { common::mmio::write32(regs::GPPUD, pull) };
    timer.delay_us(150);

    let (clk_reg, bit) = if pin < 32 {
        (regs::GPPUDCLK0, 1u32 << pin)
    } else {
        (regs::GPPUDCLK1, 1u32 << (pin - 32))
    };
    unsafe { common::mmio::write32(clk_reg, bit) };
    timer.delay_us(150);

    unsafe { common::mmio::write32(regs::GPPUD, 0) };
    unsafe { common::mmio::write32(clk_reg, 0) };
}

// ============================================================================
// Gpio trait implementation
// ============================================================================

impl Bcm2710Gpio {
    /// Configure DPI pins — GPIO 0-17 and 20-27 to ALT2, skipping 18-19 for audio.
    fn configure_dpi(&self) -> HalResult<()> {
        use hal::timer::Timer;
        let alt2 = regs::gpio_func::ALT2;

        // GPFSEL0: GPIO 0-9 all ALT2
        let gpfsel0 = (alt2) | (alt2 << 3) | (alt2 << 6) | (alt2 << 9)
                     | (alt2 << 12) | (alt2 << 15) | (alt2 << 18) | (alt2 << 21)
                     | (alt2 << 24) | (alt2 << 27);

        // GPFSEL1: GPIO 10-17 ALT2, GPIO 18-19 left as input (for PWM audio)
        let gpfsel1 = (alt2) | (alt2 << 3) | (alt2 << 6) | (alt2 << 9)
                     | (alt2 << 12) | (alt2 << 15) | (alt2 << 18) | (alt2 << 21);
        // No bits 24-29 — GPIO 18/19 reserved for audio

        // GPFSEL2: GPIO 20-27 all ALT2
        let gpfsel2 = (alt2) | (alt2 << 3) | (alt2 << 6) | (alt2 << 9)
                     | (alt2 << 12) | (alt2 << 15) | (alt2 << 18) | (alt2 << 21);

        unsafe {
            common::mmio::write32(regs::GPFSEL0, gpfsel0);
            common::mmio::write32(regs::GPFSEL1, gpfsel1);
            common::mmio::write32(regs::GPFSEL2, gpfsel2);
        }

        hal::cpu::dmb();

        // Disable pulls on DPI pins
        let timer = crate::timer::Bcm2710Timer;
        unsafe { common::mmio::write32(regs::GPPUD, regs::gpio_pull::OFF) };
        timer.delay_us(150);
        unsafe { common::mmio::write32(regs::GPPUDCLK0, 0x0FF3_FFFF) }; // bits 0-17, 20-27
        timer.delay_us(150);
        unsafe { common::mmio::write32(regs::GPPUD, 0) };
        unsafe { common::mmio::write32(regs::GPPUDCLK0, 0) };

        Ok(())
    }

    fn configure_audio(&self) -> HalResult<()> {
        use hal::timer::Timer;
        let alt5 = regs::gpio_func::ALT5;
        let timer = crate::timer::Bcm2710Timer;

        let gpfsel1 = unsafe { common::mmio::read32(regs::GPFSEL1) };
        let gpfsel1 = (gpfsel1 & 0xC0FF_FFFF) | (alt5 << 24) | (alt5 << 27);
        unsafe { common::mmio::write32(regs::GPFSEL1, gpfsel1) };

        unsafe { common::mmio::write32(regs::GPPUD, regs::gpio_pull::OFF) };
        timer.delay_us(150);
        unsafe { common::mmio::write32(regs::GPPUDCLK0, (1 << 18) | (1 << 19)) };
        timer.delay_us(150);
        unsafe { common::mmio::write32(regs::GPPUD, 0) };
        unsafe { common::mmio::write32(regs::GPPUDCLK0, 0) };

        Ok(())
    }

    fn configure_sdcard(&self) -> HalResult<()> {
        use hal::timer::Timer;
        let alt0 = regs::gpio_func::ALT0;
        let timer = crate::timer::Bcm2710Timer;

        // GPIO 48-49 in GPFSEL4 (bits 24-29)
        let gpfsel4 = unsafe { common::mmio::read32(regs::GPFSEL4) };
        let gpfsel4 = (gpfsel4 & 0xC0FF_FFFF) | (alt0 << 24) | (alt0 << 27);
        unsafe { common::mmio::write32(regs::GPFSEL4, gpfsel4) };

        // GPIO 50-53 in GPFSEL5 (bits 0-11)
        let gpfsel5 = unsafe { common::mmio::read32(regs::GPFSEL5) };
        let gpfsel5 = (gpfsel5 & 0xFFFF_F000) | alt0 | (alt0 << 3) | (alt0 << 6) | (alt0 << 9);
        unsafe { common::mmio::write32(regs::GPFSEL5, gpfsel5) };

        // Pull-ups on data/cmd lines (GPIO 49-53)
        unsafe { common::mmio::write32(regs::GPPUD, regs::gpio_pull::UP) };
        timer.delay_us(150);
        unsafe { common::mmio::write32(regs::GPPUDCLK1, (1 << 17) | (1 << 18) | (1 << 19) | (1 << 20) | (1 << 21)) };
        timer.delay_us(150);
        unsafe { common::mmio::write32(regs::GPPUD, 0) };
        unsafe { common::mmio::write32(regs::GPPUDCLK1, 0) };

        Ok(())
    }

    fn configure_uart(&self, num: u32) -> HalResult<()> {
        let timer = crate::timer::Bcm2710Timer;
        match num {
            0 => {
                set_function_raw(14, regs::gpio_func::ALT0);
                set_function_raw(15, regs::gpio_func::ALT0);
                set_pull_raw(14, regs::gpio_pull::OFF, &timer);
                set_pull_raw(15, regs::gpio_pull::UP, &timer);
                Ok(())
            }
            1 => {
                set_function_raw(14, regs::gpio_func::ALT5);
                set_function_raw(15, regs::gpio_func::ALT5);
                set_pull_raw(14, regs::gpio_pull::OFF, &timer);
                set_pull_raw(15, regs::gpio_pull::UP, &timer);
                Ok(())
            }
            _ => Err(HalError::InvalidArgument),
        }
    }

    fn configure_safe_shutdown(&self) -> HalResult<()> {
        let timer = crate::timer::Bcm2710Timer;
        set_function_raw(26, regs::gpio_func::INPUT);
        set_pull_raw(26, regs::gpio_pull::UP, &timer);
        Ok(())
    }
}

impl Gpio for Bcm2710Gpio {
    fn set_function(&self, pin: u32, func: Function) -> HalResult<()> {
        if pin > regs::GPIO_MAX_PIN { return Err(HalError::GpioInvalidPin); }
        set_function_raw(pin, hal_to_bcm(func));
        Ok(())
    }

    fn get_function(&self, pin: u32) -> HalResult<Function> {
        if pin > regs::GPIO_MAX_PIN { return Err(HalError::GpioInvalidPin); }
        Ok(bcm_to_hal(get_function_raw(pin)))
    }

    fn set_pull(&self, pin: u32, pull: Pull) -> HalResult<()> {
        if pin > regs::GPIO_MAX_PIN { return Err(HalError::GpioInvalidPin); }
        let bcm_pull = match pull {
            Pull::None => regs::gpio_pull::OFF,
            Pull::Down => regs::gpio_pull::DOWN,
            Pull::Up   => regs::gpio_pull::UP,
        };
        let timer = crate::timer::Bcm2710Timer;
        set_pull_raw(pin, bcm_pull, &timer);
        Ok(())
    }

    fn set_high(&self, pin: u32) -> HalResult<()> {
        if pin > regs::GPIO_MAX_PIN { return Err(HalError::GpioInvalidPin); }
        let (reg, bit) = if pin < 32 { (regs::GPSET0, 1u32 << pin) } else { (regs::GPSET1, 1u32 << (pin - 32)) };
        unsafe { common::mmio::write32(reg, bit) };
        Ok(())
    }

    fn set_low(&self, pin: u32) -> HalResult<()> {
        if pin > regs::GPIO_MAX_PIN { return Err(HalError::GpioInvalidPin); }
        let (reg, bit) = if pin < 32 { (regs::GPCLR0, 1u32 << pin) } else { (regs::GPCLR1, 1u32 << (pin - 32)) };
        unsafe { common::mmio::write32(reg, bit) };
        Ok(())
    }

    fn read(&self, pin: u32) -> bool {
        if pin > regs::GPIO_MAX_PIN { return false; }
        let (reg, bit) = if pin < 32 { (regs::GPLEV0, 1u32 << pin) } else { (regs::GPLEV1, 1u32 << (pin - 32)) };
        (unsafe { common::mmio::read32(reg) } & bit) != 0
    }

    fn is_valid(&self, pin: u32) -> bool { pin <= regs::GPIO_MAX_PIN }
    fn max_pin(&self) -> u32 { regs::GPIO_MAX_PIN }

    fn set_mask(&self, mask: u32, bank: u32) -> HalResult<()> {
        match bank {
            0 => { unsafe { common::mmio::write32(regs::GPSET0, mask) }; Ok(()) }
            1 => { unsafe { common::mmio::write32(regs::GPSET1, mask) }; Ok(()) }
            _ => Err(HalError::InvalidArgument),
        }
    }

    fn clear_mask(&self, mask: u32, bank: u32) -> HalResult<()> {
        match bank {
            0 => { unsafe { common::mmio::write32(regs::GPCLR0, mask) }; Ok(()) }
            1 => { unsafe { common::mmio::write32(regs::GPCLR1, mask) }; Ok(()) }
            _ => Err(HalError::InvalidArgument),
        }
    }

    fn read_mask(&self, bank: u32) -> u32 {
        match bank {
            0 => unsafe { common::mmio::read32(regs::GPLEV0) },
            1 => unsafe { common::mmio::read32(regs::GPLEV1) },
            _ => 0,
        }
    }

    fn configure_peripheral(&self, peripheral: Peripheral) -> HalResult<()> {
        match peripheral {
            Peripheral::Dpi          => self.configure_dpi(),
            Peripheral::Hdmi         => Ok(()), // No GPIO config needed on BCM2710
            Peripheral::SdCard       => self.configure_sdcard(),
            Peripheral::Audio        => self.configure_audio(),
            Peripheral::Uart(n)      => self.configure_uart(n),
            Peripheral::SafeShutdown => self.configure_safe_shutdown(),
            _ => Err(HalError::NotSupported),
        }
    }

    fn safe_shutdown_triggered(&self) -> bool {
        // GPIO 26 is active-low on GPi Case
        !self.read(26)
    }
}
